### PR TITLE
    jenkins-plugin: add retry layer/adapter for REST invocations

### PR DESCRIPTION
**Pull request by gabemontero**
Bug 1454485
    Bugzilla link https://bugzilla.redhat.com/show_bug.cgi?id=1454485
    Detailed explanation
    Implement a general retry mechanism for failing HTTP REST operations
    against the OpenShift master

@bparees @oatmealraisin FYI

### CODE REVIEW

File: src/main/java/com/openshift/jenkins/plugins/pipeline/MessageConstants.java [hunk 1]
```diff
@@ -159,4 +159,10 @@
 public static final String SOCKET_TIMEOUT = " a socket level communication timeout to \"%s\" occurred.";
 public static final String HTTP_ERR = " the HTTP level communication error \"%s\" for \"%s\" occurred.";
 
+/*
+ * These messages are centered around restclient retry
+ */
+public static final String RETRY = "\nA execption occurred invoking a REST operation against the OpenShift master.  The operation will be retried.  Exception message \"%s\".";
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/MessageConstants.java]: exception.

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            } catch (Throwable t) {
+                ose = new OpenShiftException(t, "retry", call);
+                listener.getLogger().println(String.format(MessageConstants.RETRY, t.getMessage()));
+                retryCount++;
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e) {
+                    ose = new OpenShiftException(e, "retry", call);
+                }
+            }
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java]: needs a return if we didn't catch an exception?

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    private final IClient client;
+    private final TaskListener listener;
+    
+    private Object retry(Callable<Object> call) throws OpenShiftException {
+        int retryCount = 0;
+        OpenShiftException ose = null;
+        while (retryCount < MAX_RETRY) {
+            try {
+                Object o = call.call();
+                return o;
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java:66]: I assume standard failures like "permission denied" or "unknown resource" do not result in exceptions from call()?

**Review by bparees — CHANGES_REQUESTED** [review]

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java (hunk 2 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java]: damn i had that in there at one time .... I was messing with `Void` vs. `void` and must have forgot to put it back ... that resulted in the test failure too.   Thanks for the catch.

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java (hunk 3 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/model/RetryIClient.java:66]: Another good catch ... turns out restclient has `UnauthorizedException`, `NotFoundException`, and `BadRequestException` (rc's 403, 404, and 400).  I'll check for those and *NOT* retry.

**Comment by gabemontero** [comment]
Ah ... the httpGet path needs some work

**Comment by gabemontero** [comment]
ok tests are clean now with some more tweaks to the error handling .. have 1 more mod I'm about to try

**Comment by gabemontero** [comment]
interesting ... a concurrent mod in this run uncovered a NPE ... will address NPE, and investigate if there is a deeper issue wrt retry

**Comment by gabemontero** [comment]
ah yes ... the master sends back an HTTP rc 409, which the restclient does not deal with specifically ... we already had retry for this higher up the plugin stack

IMO simply addressing the NPE in the new logic is sufficient  ... will push momentarily

**Comment by gabemontero** [comment]
OK ... tests are looking good.

@bparees - any more suggestions before merging?

**Comment by bparees** [comment]
lgtm

**Pull request merged**